File: sdk/js/packages/node/src/schema.ts
// Copyright 2022 Touca, Inc. Subject to Apache-2.0 License.

import { Builder, Offset } from 'flatbuffers';

export enum Type {
  NONE = 0,
  Bool = 1,
  Int = 2,
  UInt = 3,
  Float = 4,
  Double = 5,
  String = 6,
  Object = 7,
  Array = 8,
  Blob = 9
}

export enum ComparisonRuleMode {
  Absolute = 0,
  Relative = 1
}

export enum ResultType {
  Check = 1,
  Assert = 2
}

export class ComparisonRuleDouble {
  static startComparisonRuleDouble(builder: Builder) {
    builder.startObject(4);
  }

  static addMode(builder: Builder, mode: ComparisonRuleMode) {
    builder.addFieldInt8(0, mode, ComparisonRuleMode.Absolute);
  }

  static addMax(builder: Builder, max: number) {
    builder.addFieldFloat64(1, max, 0);
  }

  static addMin(builder: Builder, min: number) {
    builder.addFieldFloat64(2, min, 0);
  }

  static addPercent(builder: Builder, percent: boolean) {
    builder.addFieldInt8(3, +percent, 0);
  }

  static endComparisonRuleDouble(builder: Builder): Offset {
    const offset = builder.endObject();
    return offset;
  }
}

export class TypeWrapper {
  static startTypeWrapper(builder: Builder) {
    builder.startObject(2);
  }

  static addValueType(builder: Builder, valueType: Type) {
    builder.addFieldInt8(0, valueType, Type.NONE);
  }

  static addValue(builder: Builder, valueOffset: Offset) {
    builder.addFieldOffset(1, valueOffset, 0);
  }

  static endTypeWrapper(builder: Builder): Offset {
    const offset = builder.endObject();
    return offset;
  }
}

export class Bool {
  static startBool(builder: Builder) {
    builder.startObject(1);
  }

  static addValue(builder: Builder, value: boolean) {
    builder.addFieldInt8(0, +value, +false);
  }

  static endBool(builder: Builder): Offset {
    const offset = builder.endObject();
    return offset;
  }
}

export class Int {
  static startInt(builder: Builder) {
    builder.startObject(1);
  }

  static addValue(builder: Builder, value: bigint) {
    builder.addFieldInt64(0, value, BigInt('0'));
  }

  static endInt(builder: Builder): Offset {
    const offset = builder.endObject();
    return offset;
  }
}

export class Double {
  static startDouble(builder: Builder) {
    builder.startObject(2);
  }

  static addValue(builder: Builder, value: number) {
    builder.addFieldFloat64(0, value, 0.0);
  }

  static addRule(builder: Builder, ruleOffset: Offset) {
    builder.addFieldOffset(1, ruleOffset, 0);
  }

  static endDouble(builder: Builder): Offset {
    const offset = builder.endObject();
    return offset;
  }
}

export class T_String {
  static startString(builder: Builder) {
    builder.startObject(1);
  }

  static addValue(builder: Builder, valueOffset: Offset) {
    builder.addFieldOffset(0, valueOffset, 0);
  }

  static endString(builder: Builder): Offset {
    const offset = builder.endObject();
    return offset;
  }
}

export class ObjectMember {
  static startObjectMember(builder: Builder) {
    builder.startObject(2);
  }

  static addName(builder: Builder, nameOffset: Offset) {
    builder.addFieldOffset(0, nameOffset, 0);
  }

  static addValue(builder: Builder, valueOffset: Offset) {
    builder.addFieldOffset(1, valueOffset, 0);
  }

  static endObjectMember(builder: Builder): Offset {
    const offset = builder.endObject();
    return offset;
  }
}

export class T_Object {
  static startObject(builder: Builder) {
    builder.startObject(2);
  }

  static addKey(builder: Builder, keyOffset: Offset) {
    builder.addFieldOffset(0, keyOffset, 0);
  }

  static addValues(builder: Builder, valuesOffset: Offset) {
    builder.addFieldOffset(1, valuesOffset, 0);
  }

  static createValuesVector(builder: Builder, data: Offset[]): Offset {
    builder.startVector(4, data.length, 4);
    for (let i = data.length - 1; i >= 0; i--) {
      builder.addOffset(data[i]);
    }
    return builder.endVector();
  }

  static endObject(builder: Builder): Offset {
    const offset = builder.endObject();
    return offset;
  }
}

export class Array {
  static startArray(builder: Builder) {
    builder.startObject(1);
  }

  static addValues(builder: Builder, valuesOffset: Offset) {
    builder.addFieldOffset(0, valuesOffset, 0);
  }

  static createValuesVector(builder: Builder, data: Offset[]): Offset {
    builder.startVector(4, data.length, 4);
    for (let i = data.length - 1; i >= 0; i--) {
      builder.addOffset(data[i]);
    }
    return builder.endVector();
  }

  static endArray(builder: Builder): Offset {
    const offset = builder.endObject();
    return offset;
  }
}

export class Blob {
  static startBlob(builder: Builder) {
    builder.startObject(3);
  }

  static addDigest(builder: Builder, digestOffset: Offset) {
    builder.addFieldOffset(0, digestOffset, 0);
  }

  static addMimetype(builder: Builder, mimetypeOffset: Offset) {
    builder.addFieldOffset(1, mimetypeOffset, 0);
  }

  static addReference(builder: Builder, referenceOffset: Offset) {
    builder.addFieldOffset(2, referenceOffset, 0);
  }

  static endBlob(builder: Builder): Offset {
    const offset = builder.endObject();
    return offset;
  }
}

export class Result {
  static startResult(builder: Builder) {
    builder.startObject(3);
  }

  static addKey(builder: Builder, keyOffset: Offset) {
    builder.addFieldOffset(0, keyOffset, 0);
  }

  static addValue(builder: Builder, valueOffset: Offset) {
    builder.addFieldOffset(1, valueOffset, 0);
  }

  static addTyp(builder: Builder, typ: ResultType) {
    builder.addFieldInt8(2, typ, ResultType.Check);
  }

  static endResult(builder: Builder): Offset {
    const offset = builder.endObject();
    return offset;
  }
}

export class Metric {
  static startMetric(builder: Builder) {
    builder.startObject(2);
  }

  static addKey(builder: Builder, keyOffset: Offset) {
    builder.addFieldOffset(0, keyOffset, 0);
  }

  static addValue(builder: Builder, valueOffset: Offset) {
    builder.addFieldOffset(1, valueOffset, 0);
  }

  static endMetric(builder: Builder): Offset {
    const offset = builder.endObject();
    return offset;
  }
}

export class Results {
  static startResults(builder: Builder) {
    builder.startObject(1);
  }

  static addEntries(builder: Builder, entriesOffset: Offset) {
    builder.addFieldOffset(0, entriesOffset, 0);
  }

  static createEntriesVector(builder: Builder, data: Offset[]): Offset {
    builder.startVector(4, data.length, 4);
    for (let i = data.length - 1; i >= 0; i--) {
      builder.addOffset(data[i]);
    }
    return builder.endVector();
  }

  static endResults(builder: Builder): Offset {
    const offset = builder.endObject();
    return offset;
  }
}

export class Metrics {
  static startMetrics(builder: Builder) {
    builder.startObject(1);
  }

  static addEntries(builder: Builder, entriesOffset: Offset) {
    builder.addFieldOffset(0, entriesOffset, 0);
  }

  static createEntriesVector(builder: Builder, data: Offset[]): Offset {
    builder.startVector(4, data.length, 4);
    for (let i = data.length - 1; i >= 0; i--) {
      builder.addOffset(data[i]);
    }
    return builder.endVector();
  }

  static endMetrics(builder: Builder): Offset {
    const offset = builder.endObject();
    return offset;
  }
}

export class Metadata {
  static startMetadata(builder: Builder) {
    builder.startObject(6);
  }

  static addTestsuite(builder: Builder, testsuiteOffset: Offset) {
    builder.addFieldOffset(0, testsuiteOffset, 0);
  }

  static addVersion(builder: Builder, versionOffset: Offset) {
    builder.addFieldOffset(1, versionOffset, 0);
  }

  static addTestcase(builder: Builder, testcaseOffset: Offset) {
    builder.addFieldOffset(3, testcaseOffset, 0);
  }

  static addBuiltAt(builder: Builder, builtAtOffset: Offset) {
    builder.addFieldOffset(4, builtAtOffset, 0);
  }

  static addTeamslug(builder: Builder, teamslugOffset: Offset) {
    builder.addFieldOffset(5, teamslugOffset, 0);
  }

  static endMetadata(builder: Builder): Offset {
    const offset = builder.endObject();
    return offset;
  }
}

export class Message {
  static startMessage(builder: Builder) {
    builder.startObject(4);
  }

  static addMetadata(builder: Builder, metadataOffset: Offset) {
    builder.addFieldOffset(0, metadataOffset, 0);
  }

  static addResults(builder: Builder, resultsOffset: Offset) {
    builder.addFieldOffset(1, resultsOffset, 0);
  }

  static addMetrics(builder: Builder, metricsOffset: Offset) {
    builder.addFieldOffset(3, metricsOffset, 0);
  }

  static endMessage(builder: Builder): Offset {
    const offset = builder.endObject();
    return offset;
  }
}

export class MessageBuffer {
  static startMessageBuffer(builder: Builder) {
    builder.startObject(1);
  }

  static addBuf(builder: Builder, bufOffset: Offset) {
    builder.addFieldOffset(0, bufOffset, 0);
  }

  static createBufVector(
    builder: Builder,
    data: number[] | Uint8Array
  ): Offset {
    builder.startVector(1, data.length, 1);
    for (let i = data.length - 1; i >= 0; i--) {
      builder.addInt8(data[i]);
    }
    return builder.endVector();
  }

  static endMessageBuffer(builder: Builder): Offset {
    const offset = builder.endObject();
    return offset;
  }
}

export class Messages {
  static startMessages(builder: Builder) {
    builder.startObject(1);
  }

  static addMessages(builder: Builder, messagesOffset: Offset) {
    builder.addFieldOffset(0, messagesOffset, 0);
  }

  static createMessagesVector(builder: Builder, data: Offset[]): Offset {
    builder.startVector(4, data.length, 4);
    for (let i = data.length - 1; i >= 0; i--) {
      builder.addOffset(data[i]);
    }
    return builder.endVector();
  }

  static endMessages(builder: Builder): Offset {
    const offset = builder.endObject();
    return offset;
  }
}

File: sdk/java/touca/src/main/java/io/touca/core/Schema.java
// Copyright 2022 Touca, Inc. Subject to Apache-2.0 License.

package io.touca.core;

import com.google.flatbuffers.FlatBufferBuilder;
import com.google.flatbuffers.Table;

public final class Schema {

  private Schema() {
  }

  public static final class TType {
    public static final byte NONE = 0;
    public static final byte TBool = 1;
    public static final byte TInt = 2;
    public static final byte TUInt = 3;
    public static final byte TFloat = 4;
    public static final byte TDouble = 5;
    public static final byte TString = 6;
    public static final byte TObject = 7;
    public static final byte TArray = 8;
    public static final byte TBlob = 9;

    private TType() {
    }
  }

  public static final class ComparisonRuleMode {
    public static final int Absolute = 0;
    public static final int Relative = 1;

    private ComparisonRuleMode() {
    }
  }

  public static final class ResultType {
    public static final int Check = 1;
    public static final int Assert = 2;

    private ResultType() {
    }
  }

  public static final class ComparisonRuleDouble extends Table {
    public static void startComparisonRuleDouble(final FlatBufferBuilder builder) {
      builder.startTable(4);
    }

    public static void addMode(final FlatBufferBuilder builder, final int mode) {
      builder.addByte(0, (byte) mode, (byte) 0);
    }

    public static void addMax(final FlatBufferBuilder builder, final double max) {
      builder.addDouble(1, max, 0.0);
    }

    public static void addMin(final FlatBufferBuilder builder, final double min) {
      builder.addDouble(2, min, 0.0);
    }

    public static void addPercent(final FlatBufferBuilder builder, final boolean percent) {
      builder.addBoolean(3, percent, false);
    }

    public static int endComparisonRuleDouble(final FlatBufferBuilder builder) {
      return builder.endTable();
    }
  }

  public static final class TypeWrapper extends Table {

    public static void startTypeWrapper(final FlatBufferBuilder builder) {
      builder.startTable(2);
    }

    public static void addValueType(final FlatBufferBuilder builder,
        final byte valueType) {
      builder.addByte(0, valueType, 0);
    }

    public static void addValue(final FlatBufferBuilder builder,
        final int valueOffset) {
      builder.addOffset(1, valueOffset, 0);
    }

    public static int endTypeWrapper(final FlatBufferBuilder builder) {
      return builder.endTable();
    }
  }

  public static final class TBool extends Table {

    public static void startBool(final FlatBufferBuilder builder) {
      builder.startTable(1);
    }

    public static void addValue(final FlatBufferBuilder builder,
        final boolean value) {
      builder.addBoolean(0, value, false);
    }

    public static int endBool(final FlatBufferBuilder builder) {
      return builder.endTable();
    }

  }

  public static final class TInt extends Table {
    public static void startTInt(final FlatBufferBuilder builder) {
      builder.startTable(1);
    }

    public static void addValue(final FlatBufferBuilder builder,
        final long value) {
      builder.addLong(0, value, 0L);
    }

    public static int endTInt(final FlatBufferBuilder builder) {
      return builder.endTable();
    }
  }

  public static final class TDouble extends Table {

    public static void startTDouble(final FlatBufferBuilder builder) {
      builder.startTable(2);
    }

    public static void addValue(final FlatBufferBuilder builder,
        final double value) {
      builder.addDouble(0, value, 0.0);
    }

    public static void addRule(final FlatBufferBuilder builder, final int ruleOffset) {
      builder.addOffset(1, ruleOffset, 0);
    }

    public static int endTDouble(final FlatBufferBuilder builder) {
      return builder.endTable();
    }
  }

  public static final class TString extends Table {

    public static void startTString(final FlatBufferBuilder builder) {
      builder.startTable(1);
    }

    public static void addValue(final FlatBufferBuilder builder,
        final int valueOffset) {
      builder.addOffset(0, valueOffset, 0);
    }

    public static int endTString(final FlatBufferBuilder builder) {
      return builder.endTable();
    }
  }

  public static final class TObjectMember extends Table {

    public static void startTObjectMember(final FlatBufferBuilder builder) {
      builder.startTable(2);
    }

    public static void addName(final FlatBufferBuilder builder,
        final int nameOffset) {
      builder.addOffset(0, nameOffset, 0);
    }

    public static void addValue(final FlatBufferBuilder builder,
        final int valueOffset) {
      builder.addOffset(1, valueOffset, 0);
    }

    public static int endTObjectMember(final FlatBufferBuilder builder) {
      return builder.endTable();
    }
  }

  public static final class TObject extends Table {

    public static void startTObject(final FlatBufferBuilder builder) {
      builder.startTable(2);
    }

    public static void addKey(final FlatBufferBuilder builder,
        final int keyOffset) {
      builder.addOffset(0, keyOffset, 0);
    }

    public static void addValues(final FlatBufferBuilder builder,
        final int valuesOffset) {
      builder.addOffset(1, valuesOffset, 0);
    }

    public static int createValuesVector(final FlatBufferBuilder builder,
        final int[] data) {
      builder.startVector(4, data.length, 4);
      for (int i = data.length - 1; i >= 0; i--)
        builder.addOffset(data[i]);
      return builder.endVector();
    }

    public static int endTObject(final FlatBufferBuilder builder) {
      return builder.endTable();
    }
  }

  public static final class TArray extends Table {

    public static void startTArray(final FlatBufferBuilder builder) {
      builder.startTable(1);
    }

    public static void addValues(final FlatBufferBuilder builder,
        final int valuesOffset) {
      builder.addOffset(0, valuesOffset, 0);
    }

    public static int createValuesVector(final FlatBufferBuilder builder,
        final int[] data) {
      builder.startVector(4, data.length, 4);
      for (int i = data.length - 1; i >= 0; i--)
        builder.addOffset(data[i]);
      return builder.endVector();
    }

    public static int endTArray(final FlatBufferBuilder builder) {
      return builder.endTable();
    }
  }

  public static final class TBlob extends Table {
    public static void startTBlob(final FlatBufferBuilder builder) {
      builder.startTable(3);
    }

    public static void addDigest(final FlatBufferBuilder builder,
        final int digestOffset) {
      builder.addOffset(0, digestOffset, 0);
    }

    public static void addMimetype(final FlatBufferBuilder builder,
        final int mimetypeOffset) {
      builder.addOffset(1, mimetypeOffset, 0);
    }

    public static void addReference(final FlatBufferBuilder builder,
        final int referenceOffset) {
      builder.addOffset(2, referenceOffset, 0);
    }

    public static int endTBlob(final FlatBufferBuilder builder) {
      return builder.endTable();
    }
  }

  public static final class Result extends Table {

    public static void startResult(final FlatBufferBuilder builder) {
      builder.startTable(3);
    }

    public static void addKey(final FlatBufferBuilder builder,
        final int keyOffset) {
      builder.addOffset(0, keyOffset, 0);
    }

    public static void addValue(final FlatBufferBuilder builder,
        final int valueOffset) {
      builder.addOffset(1, valueOffset, 0);
    }

    public static void addTyp(final FlatBufferBuilder builder, final int typ) {
      builder.addByte(2, (byte) typ, (byte) 1);
    }

    public static int endResult(final FlatBufferBuilder builder) {
      return builder.endTable();
    }
  }

  public static final class Metric extends Table {

    public static void startMetric(final FlatBufferBuilder builder) {
      builder.startTable(2);
    }

    public static void addKey(final FlatBufferBuilder builder,
        final int keyOffset) {
      builder.addOffset(0, keyOffset, 0);
    }

    public static void addValue(final FlatBufferBuilder builder,
        final int valueOffset) {
      builder.addOffset(1, valueOffset, 0);
    }

    public static int endMetric(final FlatBufferBuilder builder) {
      return builder.endTable();
    }
  }

  public static final class Results extends Table {

    public static void startResults(final FlatBufferBuilder builder) {
      builder.startTable(1);
    }

    public static void addEntries(final FlatBufferBuilder builder,
        final int entriesOffset) {
      builder.addOffset(0, entriesOffset, 0);
    }

    public static int createEntriesVector(final FlatBufferBuilder builder,
        final int[] data) {
      builder.startVector(4, data.length, 4);
      for (int i = data.length - 1; i >= 0; i--)
        builder.addOffset(data[i]);
      return builder.endVector();
    }

    public static int endResults(final FlatBufferBuilder builder) {
      return builder.endTable();
    }
  }

  public static final class Metrics extends Table {

    public static void startMetrics(final FlatBufferBuilder builder) {
      builder.startTable(1);
    }

    public static void addEntries(final FlatBufferBuilder builder,
        int entriesOffset) {
      builder.addOffset(0, entriesOffset, 0);
    }

    public static int createEntriesVector(final FlatBufferBuilder builder,
        final int[] data) {
      builder.startVector(4, data.length, 4);
      for (int i = data.length - 1; i >= 0; i--)
        builder.addOffset(data[i]);
      return builder.endVector();
    }

    public static int endMetrics(final FlatBufferBuilder builder) {
      return builder.endTable();
    }
  }

  public static final class Metadata extends Table {

    public static void startMetadata(final FlatBufferBuilder builder) {
      builder.startTable(6);
    }

    public static void addTestsuite(final FlatBufferBuilder builder,
        final int testsuiteOffset) {
      builder.addOffset(0, testsuiteOffset, 0);
    }

    public static void addVersion(final FlatBufferBuilder builder,
        final int versionOffset) {
      builder.addOffset(1, versionOffset, 0);
    }

    public static void addTestcase(final FlatBufferBuilder builder,
        final int testcaseOffset) {
      builder.addOffset(3, testcaseOffset, 0);
    }

    public static void addBuiltAt(final FlatBufferBuilder builder,
        final int builtAtOffset) {
      builder.addOffset(4, builtAtOffset, 0);
    }

    public static void addTeamslug(final FlatBufferBuilder builder,
        final int teamslugOffset) {
      builder.addOffset(5, teamslugOffset, 0);
    }

    public static int endMetadata(final FlatBufferBuilder builder) {
      return builder.endTable();
    }
  }

  public static final class Message extends Table {

    public static void startMessage(final FlatBufferBuilder builder) {
      builder.startTable(4);
    }

    public static void addMetadata(final FlatBufferBuilder builder,
        final int metadataOffset) {
      builder.addOffset(0, metadataOffset, 0);
    }

    public static void addResults(final FlatBufferBuilder builder,
        final int resultsOffset) {
      builder.addOffset(1, resultsOffset, 0);
    }

    public static void addMetrics(final FlatBufferBuilder builder,
        final int metricsOffset) {
      builder.addOffset(3, metricsOffset, 0);
    }

    public static int endMessage(final FlatBufferBuilder builder) {
      return builder.endTable();
    }
  }

  public static final class MessageBuffer extends Table {

    public static void startMessageBuffer(final FlatBufferBuilder builder) {
      builder.startTable(1);
    }

    public static void addBuf(final FlatBufferBuilder builder,
        final int bufOffset) {
      builder.addOffset(0, bufOffset, 0);
    }

    public static int createBufVector(final FlatBufferBuilder builder,
        final byte[] data) {
      return builder.createByteVector(data);
    }

    public static int endMessageBuffer(final FlatBufferBuilder builder) {
      return builder.endTable();
    }
  }

  public static final class Messages extends Table {
    public static void startMessages(final FlatBufferBuilder builder) {
      builder.startTable(1);
    }

    public static void addMessages(final FlatBufferBuilder builder,
        final int messagesOffset) {
      builder.addOffset(0, messagesOffset, 0);
    }

    public static int createMessagesVector(final FlatBufferBuilder builder,
        final int[] data) {
      builder.startVector(4, data.length, 4);
      for (int i = data.length - 1; i >= 0; i--)
        builder.addOffset(data[i]);
      return builder.endVector();
    }

    public static int endMessages(final FlatBufferBuilder builder) {
      return builder.endTable();
    }
  }

}
